lib/spack/spack/environment/environment.py

```diff
@@ -422,25 +422,105 @@ def _eval_conditional(string):
-
-    _, record = spack.store.STORE.db.query_by_spec_hash(spec.dag_hash())
+    db = _database or spack.store.STORE.db
+    dev_path_var = spec.variants.get("dev_path")
+    _, record = db.query_by_spec_hash(spec.dag_hash())
     mtime = fs.last_modification_time_recursive(dev_path_var.value)
     return mtime > record.installation_time
 
 
+class GitRepoChangeDetector:
+    """Associates with a Git repository. Uses Git to determine when
+    tracked files in the repository have changed, regardless of whether
+    they have been committed or staged (i.e. with ``git add``).
+
+    This does not look for changes in untracked files, and maintains
+    extra files for bookkeeping in this directory, within a ``.spack``
+    subdirectory (creating it if it does not exist).
+    """
+
+    CHANGED = 0
+    NO_PRIOR = 1
+    NOT_CHANGED = 2
+
+    def __init__(self, dev_path):
+        self.base_dir = pathlib.Path(dev_path)
+        self.git_dir = self.base_dir / ".git"
+        self.spack_state = self.base_dir / ".spack"
+        self.cache_state = self.spack_state / "spackdev-git-hash"
+        self.current_hash = None
+
+    @staticmethod
+    def from_src_dir(dev_path):
+        git_dir = pathlib.Path(dev_path) / ".git"
+        if not git_dir.is_dir():
+            return
+        return GitRepoChangeDetector(dev_path)
+
+    def git_modification_hash(self):
+        """
+        Given a Git repository directory, output a hash that represents the current
+        state of the directory, including all uncommitted changes to tracked files
+        in the repository.
+        """
+        git_index = self.git_dir / "index"
+        tmp_index = self.spack_state / "spack-package-git-index"
+
+        try:
+            os.remove(tmp_index)
+        except FileNotFoundError:
+            pass
+        self.spack_state.mkdir(exist_ok=True)
+        shutil.copyfile(git_index, tmp_index)
+
+        env = {"GIT_INDEX_FILE": str(tmp_index)}
+        git = spack.util.git.git(required=True)
+        with fs.working_dir(self.base_dir):
+            git("add", "-u", env=env)
+            hash = git("write-tree", env=env, output=str)
+        return hash
+
+    def update_current(self):
+        """
+        Determine the current state of the Git repository (calculate
+        the hash) and return whether it differs from the prior state.
+
+        Returns:
+            (int): NO_PRIOR if there is no prior state available to
+                check; CHANGED if the new Git hash differs, and
+                NOT_CHANGED otherwise.
+        """
+        self.current_hash = self.git_modification_hash()
+        prior_hash = self.prior_hash()
+        if not prior_hash:
+            # Note: originally I thought about returning NOT_CHANGED
+            # if the git state is "clean", but technically a user
+            # could have initially installed with a dirty state, and
+            # then reverted their changes.
+            return GitRepoChangeDetector.NO_PRIOR
```

Comment:
is there a good reason not to just say that this is `CHANGED`?  They're likely to need to build the first time anyway.

Reply:
For users who were depending on timestamp detection, this allows for them to use timestamps the first time after pulling this (and also calculate the Git state, so the next time, it can just use Git)

Reply:
I'm saying why not build the first time, record that it was built, and calculate the git hash at that time, rather than relying on timestamps?

Reply:
We could do that, but to clarify: if we do it that way, we may possibly rebuild unnecessarily the first time we run this logic after pulling the new feature. I decided it would be worthwhile to detect that case and fall back on timestamps, but it sounds like you don't see that as particularly useful. 

IMO there's no issue with forcing a rebuild the first time to simplify the logic, but this assumes that someone would in fact find it simpler for us to default to indicating there was a change when we don't have a prior (which IMO is debatable).

Anyhow, I'll assume that I should remove the `NO_PRIOR` logic.

Reply:
Is there a reason the git logic *has* to be unaware of untracked files?  We could use `git status --porcelain` to see even unknown files in the directory, e.g. in my spack repo right now:

```console
> git status --porcelain
 M lib/spack/spack/traverse.py
?? check_all_dicts.py
?? non-const
?? nonconst.py
```

One modified file and 3 unknowns -- pretty easy to parse, and `git status` is probably the fastest way to find this out. You could add the unknowns to the index, which would include them in the tree hash.

Also, it seems like we could use this to avoid a full traversal of the directory. You know that `git status` reports only what is different from the last commit, and you know the modification date of the last commit.  if we find that we have no prior, couldn't we avoid a full traversal by looking at the last commit and the modification date of any files mentioned by `git status`? You can terminate early if you find anything newer than the last build of the package.

Reply:
(this appears to be a response to https://github.com/spack/spack/pull/41013#discussion_r1430695886, so I'll respond with that in mind)

> Is there a reason the git logic has to be unaware of untracked files? 

It does not *have* to be unaware, but my guess is most of the speedup over timestamp checking is achieved by not looking for untracked files.

> One modified file and 3 unknowns -- pretty easy to parse, 

Parsing is not my concern. I think `git status` has to do a traversal, and I think that traversal is costly. For example, running `git status` on Spack's directory in an NFS dir yields:

```
$ git status
...
It took 13.95 seconds to enumerate untracked files. 'status -uno'
may speed it up, but you have to be careful not to forget to add
new files yourself (see 'git help status').
```

FWIW, repeated invocations of `git status` in this context take much less time - only ~3 seconds or so - although I'm not sure if that's on account of FS caching or something smart that Git is doing.

> You know that git status reports only what is different from the last commit, and you know the modification date of the last commit. if we find that we have no prior, couldn't we avoid a full traversal by looking at the last commit and the modification date of any files mentioned by git status?

I don't know how `git status` itself finds untracked files other than traversing directories, e.g.

```
git status
touch a/file/deep/within/the/directory/hierarchy
git status # how would Git know where to look?
```

FWIW I assume that whenever Git has to do traversals, it does that as quickly as it can (which is to say, even if we did force `git status` it would still probably be faster than timestamp checking).

> You could add the unknowns to the index, which would include them in the tree hash.

Also note that without something like https://github.com/spack/spack/pull/41373, build artifacts would appear as untracked files in the git repo (that PR was made with other intentions, but it would also ease this).

> You can terminate early if you find anything newer than the last build of the package.

For single-file-change updates, that would on average cut the cost in half. When developing several packages that typically do not change, this still incurs the traversal cost each time for those.

Reply:
> FWIW, repeated invocations of git status in this context take much less time - only ~3 seconds or so - although I'm not sure if that's on account of FS caching or something smart that Git is doing.

Yep that's FS caching.  What does a cold `git status` look like on a repo with `feature.manyFiles=true`?  Is it faster?

> FWIW I assume that whenever Git has to do traversals, it does that as quickly as it can (which is to say, even if we did force git status it would still probably be faster than timestamp checking).

Yes.

> For single-file-change updates, that would on average cut the cost in half. When developing several packages that typically do not change, this still incurs the traversal cost each time for those.

I don't think this is completely true, because you're speeding up the traversal just by delegating it to `git`.  One of the issues here is that python is so slow.  Does a Python traversal take 13 seconds?  Or more?

How fast is doing the tree hash?  I don't see how `git add -u` is going to be too much faster than `git status`, simply because *it* has to look at tracked files.  That's not necessarily a directory traversal, but it does require `git` to stat the files to see what tracked files have changed.